File: app/src/main/java/com/example/liftoff/command/PullCommand.java
package com.example.liftoff.command;

import com.example.liftoff.extractor.email.EmailExtractor;
import com.example.liftoff.extractor.email.GmailExtractor;
import com.example.liftoff.extractor.email.ImapGmailExtractor;
import com.example.liftoff.extractor.email.OutlookExtractor;
import com.example.liftoff.storage.Storage;
import com.example.liftoff.storage.filesystem.FileStorage;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;

/**
 * Implements a ${@code Command} for the email attachments download
 */
class PullCommand implements Command {
    /**
     * An address from which the attachments will be downloaded
     */
    private final String emailAddress;
    /**
     * A token needed to access the email service API (Gmail, Graph, etc.).
     * TODO: Move this to the properties file
     */
    private final String userToken;
    /**
     * An optional directory path for the downloaded attachments
     */
    private final Optional<String> rootDirectory;

    private static String DEFAULT_DESTINATION_DIRECTORY = "email_attachments_downloads";

    public PullCommand(String emailAddress, String userToken, Optional<String> rootDirectory) {
        this.emailAddress = emailAddress;
        this.userToken = userToken;
        this.rootDirectory = rootDirectory;
    }

    /**
     * Check if the provided {@code destinationDirectory} exists in the file system.
     *
     * @return ${@code true} if {@code destinationDirectory} exists, {@code false} otherwise
     * TODO: Check for the {@code emailAddress} validity
     */
    @Override
    public boolean isValid() {
        var isValid = true;
        if (this.rootDirectory.isPresent() && !Files.exists(Paths.get(this.rootDirectory.get()))) {
            System.err.println("Provided destination path " + this.rootDirectory.get() + " for the downloaded attachments does not exist.\nPlease make sure that the provided path is valid and exists in the file system");
            isValid = false;
        }
        return isValid;
    }

    /**
     * Returns the specific ${@code EmailExtractor} implementation,
     * depending on the email adrress provided to the application
     */
    private EmailExtractor chooseSpecificEmailExtractor(final Storage fileStorage) {
        EmailExtractor extractor;
        if (this.emailAddress.endsWith("@gmail.com")) {
            extractor = new GmailExtractor(this.emailAddress, this.userToken, fileStorage);
        } else if (emailAddress.endsWith("@outlook.com")) {
            extractor = new OutlookExtractor(this.emailAddress, this.userToken, fileStorage);
        }
        // TODO: Since only ImapGmailAPI is implemented, fallback to it as a default. Remove after implementing GmailExtractor and OutlookExtractor
        extractor = new ImapGmailExtractor(this.emailAddress, this.userToken, fileStorage);
        return extractor;
    }

    /**
     * Returns the specific ${@code Storage} implementation,
     * depending on the cli arguments provided
     */
    private Storage chooseStorage(final Path rootPath) {
        Storage storage;
        // TODO instead of returning the default FileStorage implementation, add logic to choose between file system or various web interfaces (GoogleDrive, Dropbox, etc.)
        storage = new FileStorage(rootPath.toAbsolutePath());
        return storage;
    }

    /**
     * Constructs an appropriate extractor object (eg. ${code ImapGmailExtractor}) and
     * a storage object (eg. ${code FileStorage}) for the successful email attachments download
     */
    @Override
    public void execute() {
        final var rootPath = this.rootDirectory
                .map(s -> Paths.get(s).toAbsolutePath())
                .orElseGet(() -> Paths.get("", DEFAULT_DESTINATION_DIRECTORY));
        System.out.println("Saving extracted email attachments to: " + rootPath.toAbsolutePath());
        final var storage = chooseStorage(rootPath);
        final var emailExtractor = this.chooseSpecificEmailExtractor(storage);
        try {
            emailExtractor.authenticate();
            emailExtractor.extract();
            emailExtractor.close();
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
    }

}

File: src/main/scala/com/example/liftoff/command/PullCommand.scala
package com.example.liftoff.command

import com.example.liftoff.extractor.email.EmailExtractor
import com.example.liftoff.extractor.email.gmail.GmailExtractor
import com.example.liftoff.storage.Storage
import com.example.liftoff.storage.filesystem.FileStorage
import zio.{Console, UIO, ZIO, ZLayer}

import java.nio.file.{Files, Path, Paths}

class PullCommand(email: String, credentialPath: String, rootDirectory: Option[String]) extends Command {
  private val defaultDestinationDirectory = "email_attachments_downloads"

  private def rootDirectoryErrorMessage(path: String): String =
    s"Provided destination path $path for the downloaded attachments does not exist.\nPlease make sure that the provided path is valid and exists in the file system"

  private def credentialFileErrorMessage(path: String): String =
    s"Provided credential path $path does not exist.\nPlease make sure that the provided path is valid and exists in the file system"

  override def isValid: UIO[Boolean] = {
    for {
      rootDirectoryExists <- ZIO.fromOption(this.rootDirectory).foldZIO(
        _ => ZIO.succeed(false),
        path => ZIO.attempt(Files.exists(Paths.get(path))).foldZIO(
          _ => ZIO.succeed(false),
          exists => ZIO.succeed(exists)
        )
      )
      credentialFileExists <- ZIO.attempt(Files.exists(Paths.get(credentialPath))).foldZIO(
        _ => ZIO.succeed(false),
        exists => ZIO.succeed(exists)
      )
      _ <- ZIO.when(!rootDirectoryExists)(Console.printError(this.rootDirectoryErrorMessage(this.rootDirectory.getOrElse(""))).ignore)
      _ <- ZIO.when(!credentialFileExists)(Console.printError(this.credentialFileErrorMessage(credentialPath)).ignore)
      isValid <- ZIO.succeed(rootDirectoryExists && credentialFileExists)
    } yield isValid
  }

  private def chooseSpecificEmailExtractor(storage: Storage): UIO[EmailExtractor] = this.email.split("@").last match {
    case "gmail.com" => ZIO.succeed(new GmailExtractor(this.email, this.credentialPath, storage))
    //    case "outlook.com" => new OutlookExtractor(this.email, this.userToken, fileStorage)
    //    case _ => new ImapGmailExtractor(this.email, this.userToken, fileStorage)
  }

  private def chooseStorage(rootPath: Path): UIO[Storage] = {
    ZIO.succeed(new FileStorage(rootPath.toAbsolutePath))
  }

  override def execute: UIO[Unit] = {
    for {
      storagePath <- ZIO.fromOption(this.rootDirectory).foldZIO(
        _ => ZIO.succeed(Paths.get("", this.defaultDestinationDirectory)),
        path => ZIO.succeed(Paths.get(path).toAbsolutePath)
      )
      _ <- Console.printLine(s"Saving extracted email attachments to: $storagePath").ignore
      storage <- this.chooseStorage(storagePath)
      emailExtractor <- this.chooseSpecificEmailExtractor(storage)
      _ <- emailExtractor.authenticate().catchAll(error => Console.printError(error.message).ignore)
      _ <- emailExtractor.extract().catchAll(error => Console.printError(error.message).ignore)
      _ <- emailExtractor.close().catchAll(error => Console.printError(error.message).ignore)
    } yield ()
  }
}
